File: ai-service/app/config/registry.py
"""
app/config/registry.py

Loads shared-config/ - the SAME files the Backend loads (Sections 7.6-7.8,
24.2). Both services must agree on what "fishing" or "ta" or "m/s" means;
neither owns the list.

Loaded once at import. A malformed config file kills the process here rather
than producing subtly wrong behaviour later.
"""

import json
import os
import sys
from typing import Any, Dict, List, Optional

from app.config.settings import settings


def _load(filename: str) -> Dict[str, Any]:
    path = os.path.join(settings.SHARED_CONFIG_DIR, filename)
    try:
        with open(path, "r", encoding="utf-8") as fh:
            return json.load(fh)
    except Exception as exc:  # noqa: BLE001 - we want the real reason printed
        print(f"[registry] FATAL: could not load {path}: {exc}", file=sys.stderr)
        sys.exit(1)


_activities = _load("activities.json")["activities"]
_vessel_types = _load("vessel-types.json")["vessel_types"]
_languages = _load("languages.json")["languages"]
_canonical_units = _load("canonical-units.json")["units"]

ACTIVITY_IDS = {a["id"] for a in _activities}
VESSEL_TYPE_IDS = {v["id"] for v in _vessel_types}
LANGUAGE_CODES = {l["code"] for l in _languages}

_VESSEL_BY_ID = {v["id"]: v for v in _vessel_types}


def is_valid_activity(value: Optional[str]) -> bool:
    return value in ACTIVITY_IDS


def is_valid_vessel_type(value: Optional[str]) -> bool:
    return value in VESSEL_TYPE_IDS


def is_valid_language(value: Optional[str]) -> bool:
    return value in LANGUAGE_CODES


def canonical_unit(parameter: str) -> Optional[str]:
    """Return the required unit for a canonical parameter name, or None.

    None means "unknown parameter" and callers must treat that as a problem -
    never as "any unit is acceptable" (Section 24.2).
    """
    return _canonical_units.get(parameter)


def most_conservative_vessel_type() -> str:
    """Section 7.7: when vessel_type is absent for a safety request, the
    Planner substitutes the MOST conservative profile and states the assumption.
    conservatism_rank 1 = most fragile craft."""
    return min(_vessel_types, key=lambda v: v["conservatism_rank"])["id"]


def vessel_conservatism_rank(vessel_type: Optional[str]) -> int:
    """Higher rank = sturdier vessel. Unknown/missing returns the most
    conservative rank so a bad value never makes conditions look safer."""
    v = _VESSEL_BY_ID.get(vessel_type or "")
    return v["conservatism_rank"] if v else 1


ACTIVITIES: List[Dict[str, Any]] = _activities
VESSEL_TYPES: List[Dict[str, Any]] = _vessel_types
LANGUAGES: List[Dict[str, Any]] = _languages
CANONICAL_UNITS: Dict[str, str] = _canonical_units

File: backend/src/config/registry.js
/**
 * @fileoverview Shared Domain Registry and Configuration Catalog
 * @module config/registry
 * @description
 * Sections 7.6, 7.7, and 7.8:
 * Loads static canonical lists of allowed marine activities, vessel types,
 * spoken languages, and physical measurement units from the shared repository
 * `/shared-config` catalog at startup.
 *
 * Design Guarantees:
 * - Synchronization: Keeps Node.js backend and Python AI service aligned on
 *   valid IDs and semantics without code duplication.
 * - Constant-Time O(1) Validation: Pre-populates JavaScript Set and Map data
 *   structures during require-time initialization for instant request validation.
 * - Fail-Fast Loading: If any configuration JSON is unreadable or malformed,
 *   initialization throws synchronously to prevent running with partial state.
 */

const fs = require('fs');
const path = require('path');

// shared-config sits at the repo root, two levels above backend/src/config
const SHARED_CONFIG_DIR = path.resolve(__dirname, '../../../shared-config');

/**
 * Reads and parses a JSON file from the shared-config directory synchronously.
 *
 * @param {string} filename - Base name of the JSON file (e.g. 'activities.json').
 * @returns {object} Parsed JSON content.
 * @throws {Error} If the file does not exist or contains invalid JSON syntax.
 */
function loadJson(filename) {
  const filePath = path.join(SHARED_CONFIG_DIR, filename);
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  } catch (err) {
    throw new Error(
      `[registry] Could not load ${filename} from ${SHARED_CONFIG_DIR}: ${err.message}`
    );
  }
}

const activitiesFile = loadJson('activities.json');
const vesselTypesFile = loadJson('vessel-types.json');
const languagesFile = loadJson('languages.json');
const canonicalUnitsFile = loadJson('canonical-units.json');

// Build Sets/Maps once so validation is O(1) per request instead of scanning
// an array on every call.
const activityIds = new Set(activitiesFile.activities.map((a) => a.id));
const vesselTypeIds = new Set(vesselTypesFile.vessel_types.map((v) => v.id));
const languageCodes = new Set(languagesFile.languages.map((l) => l.code));

const vesselTypeById = new Map(vesselTypesFile.vessel_types.map((v) => [v.id, v]));
const activityById = new Map(activitiesFile.activities.map((a) => [a.id, a]));
const languageByCode = new Map(languagesFile.languages.map((l) => [l.code, l]));

const registry = Object.freeze({
  // ---- Raw lists (for GET endpoints that expose options to the Frontend) --
  activities: activitiesFile.activities,
  vesselTypes: vesselTypesFile.vessel_types,
  languages: languagesFile.languages,
  canonicalUnits: canonicalUnitsFile.units,

  // ---- Validation helpers (Section 7.6 / 7.7 / 7.8) ----------------------
  isValidActivity: (id) => activityIds.has(id),
  isValidVesselType: (id) => vesselTypeIds.has(id),
  isValidLanguage: (code) => languageCodes.has(code),

  // ---- Lookups -----------------------------------------------------------
  getActivity: (id) => activityById.get(id) || null,
  getVesselType: (id) => vesselTypeById.get(id) || null,
  getLanguage: (code) => languageByCode.get(code) || null,

  // Section 24.2: canonical unit for a parameter, or null if the parameter is
  // unknown. Callers MUST treat null as "cannot verify" - never as "any unit
  // is fine".
  getCanonicalUnit: (parameter) => canonicalUnitsFile.units[parameter] || null,

  // ---- Sorted valid-value lists for error messages ------------------------
  // Section 7.6: "An unknown activity value returns 400." A 400 is far more
  // useful when it tells the caller what the accepted values actually are.
  validActivityIds: () => [...activityIds].sort(),
  validVesselTypeIds: () => [...vesselTypeIds].sort(),
  validLanguageCodes: () => [...languageCodes].sort(),
});

console.log(
  `[registry] Loaded ${registry.activities.length} activities, ` +
  `${registry.vesselTypes.length} vessel types, ` +
  `${registry.languages.length} languages, ` +
  `${Object.keys(registry.canonicalUnits).length} canonical units`
);

module.exports = registry;
